### ISSUE TITLE
TypedDict with optional keys errors despite guarded access

### ISSUE THREAD

**Issue by Rapptz** [microsoft/pyright#1739]
**Describe the bug**
When using `TypedDict` with `total=False`, guarded access with `try except` reports an error.

**To Reproduce**
```py
from typing import TypedDict
class A(TypedDict, total=False):
    test: str

def get_test(a: A) -> str:
    try:
        value = a['test'] # error here
    except KeyError:
        return 'ok'
    else:
        return f'received: {value}'
```

**Expected behavior**
No error since the code is being guarded by a try-except.

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/1695103/113839398-a4f1e100-975d-11eb-808e-742945344570.png)

**VS Code extension or command-line**
VSCode. Pylance 2021.3.4.

**Comment by erictraut** [comment]
Pyright doesn't support exception-based guards. If you want this to type check, you should use a regular (non-exception-based) check:
```python
def get_test(a: A) -> str:
    if 'test' not in a:
        return 'ok'
    value = a['test']
    return f'received: {value}'
```

**Comment by Rapptz** [comment]
The actual code is a bit more complicated, I had kind of shown it a little bit with the value being used in the found case but not in the exception case.

In Python things are usually done using [EAFP rather than LBYL](https://docs.python.org/3/glossary.html#term-eafp). Likewise the behaviour of the two aren't equivalent, in the try except case you can avoid certain high cost evaluations whereas the `get` or `setdefault` cases would always pay the price for the default parameter.

Ultimately I can just use `# type: ignore` but I don't think it looks good to do this on idiomatic Python.

**Comment by erictraut** [comment]
Yeah, that's a reasonable point.

While I think it's a bad idea to use exceptions for normal code flow, I unfortunately do see this commonly in Python code. It's especially bad if you're interested in the benefits of type checking because it harms the type checker's ability to understand the code flow. Because of this, you're going to mask bugs that would otherwise be caught by the tools. It's also just harder to read and maintain, IMO.

I verified that mypy suppresses this check when it's within a try block, so I've changed pyright to do the same.

This will be in the next release.

**Comment by erictraut** [comment]
This is included in Pyright 1.1.130, which I just published. It will also be part of the next release of Pylance.

**Comment by apodznoev** [comment]
Just for the same snippet as originally posted there is the same error:
```
error: Could not access item in TypedDict
    "test" is not a required key in "A", so access may result in runtime exception (reportTypedDictNotRequiredAccess)
```
starting from 1.1.360 onwards (up to 1.1.366).

The issue was not there in 1.1.359.

**Comment by erictraut** [comment]
@apodznoev, yes this change was intentional. See [this issue](https://github.com/microsoft/pyright/issues/7714) for details. I decided to reverse the original decision because it was inconsistent with the general philosophy of type checking. Code within a `try` block should not be exempt from type checks.

If you're interested in making your code type safe, do not use exception handling for expected code flow. Instead, use regular `if` statements to check for conditions like the presence of specific keys.

**Comment by Rapptz** [comment]
That's an unfortunate change considering how overwhelmingly common try-except blocks are for accessing keys in a dictionary or other type of EAFP.

**Comment by erictraut** [comment]
I think EAFP is a very unfortunate and ill-advised practice. If you're interested in type safety and code robustness, I recommend avoiding this pattern. Exceptions should be used for exceptional cases, not anticipated code flow.

**Comment by Rapptz** [comment]
The type checker for a language should complement the predominant style of the programming language it's type checking for. I don't really want to discuss the merits of EAFP vs LBYL here but the choice is often there and commonly taken to prevent bugs like TOCTOU or for performance reasons. I understand that this is not your preferred programming style but this is honestly very common for Python. Even the notion of "exceptional cases only" for exceptions doesn't hold true here, exceptions are used all over the language for control flow internally such as `StopIteration` and the various other related exceptions when working with coroutines.

**Comment by dd-ssc** [comment]
There is also the issue of expressing the intent of the code in a (IMO) much more readable fashion with nested `TypedDicts`:

```python
response = <some nested TypedDict returned by a boto3 call into the AWS cloud>
```

EAFP:

```python
try:
    http_url_to_repo = response['repositoryMetadata']['cloneUrlHttp']    # reportTypedDictNotRequiredAccess
    local_path       = response['repositoryMetadata']['repositoryName']  # reportTypedDictNotRequiredAccess
except KeyError:
    # ... (exception handling) ...
```

LBYL:

```python
if 'repositoryMetadata' in response and \
   'cloneUrlHttp'       in response['repositoryMetadata'] and \
   'repositoryName'     in response['repositoryMetadata']:
    http_url_to_repo = response['repositoryMetadata']['cloneUrlHttp']
    local_path       = response['repositoryMetadata']['repositoryName']
else:
    # ... (error handling) ...
```

EFAP allows me to express in very few lines that the attributes are not guaranteed to exist and the code handles that; 2 pyright issues.

LBYL is free of pyright issues, but requires much more code - which IMO makes it hard to understand what the code is doing and why there are so many attribute accesses.

This obviously gets only worse with every level of nesting and every attribute - whereas the EAFP guard stays the same.

Just my 2 cents 🤷🏻‍♂️

**Comment by JimDabell** [comment]
I just hit this. Pyright should not be pushing people to rewrite idiomatic Python code into more difficult to read non-idiomatic Python code. I understand that you think this Python idiom is not the best, but Pyright is a type checker not an opinionated linter. Please reconsider. This is not going to make me rewrite a tonne of idiomatic Python, it is going to make me disable the rule.
